### ISSUE TITLE
Growing the stylelint team

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#2866]
I think we should try to grow the stylelint team. Over the past two years, the number of stylelint users has grown significantly. The number of active stylelint team members has, however, remained pretty constant. Also, for most of the team, myself included, we don't have as much time as we used to to give to the project. There simply isn't enough time amongst us to review all the PRs and triage all the issues anymore.

I think growing the stylelint team would be healthy and would help the sustainability of the project.

There are a couple of contributors I can think of off-hand:

- @CAYdenberg [has made](https://github.com/stylelint/stylelint/pulls?q=is%3Apr+author%3ACAYdenberg+is%3Aclosed) a number of contributions over the past few months.
- @modosc [has offered, and has started,](https://github.com/stylelint/stylelint/issues/2829) to port some of the stylefmt fixes over as he has experience with the stylefmt code base.
- @gucong3000 [has been active](https://github.com/stylelint/stylelint/pulls?utf8=%E2%9C%93&q=author%3Agucong3000%20) in both the stylelint and stylefmt.

@gucong3000 @CAYdenberg & @modosc Would becoming stylelint members be something you're interested in?

As team members:

1. You'll be able to help review pull requests, merge them and update the CHANGELOG.
2. You'll be able to help triage issues.
3. You'll be able to help put out releases.
4. You'll be able to help decide the direction that stylelint takes going forward (see: [VISION.md](https://github.com/stylelint/stylelint/blob/master/VISION.md)).

(I'll try to knock-up, in the next couple of days, additions to the developer guide to formalise the first two of these. We have an informal approach to issues and PRs that isn't documented anywhere).

I'm sure there are a few more suitable people I've failed to mention. Shout out if anyone else is interested! Either because you wish to support a tool you use, or as an opportunity to get involved in a open source project. As we say in the [developer guide](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md):

> If you like stylelint and open source software (since you're reading this, you almost certainly do), please consider taking some time to pitch in. Not only will you help stylelint thrive, you will also learn a thing or two — about CSS, PostCSS, Node, ES2015, unit testing, open source software, and more.

Also, are there any companies out there that use stylelint as an integral part of their tool chain and are able to spare some of their developers' time? If so, please let us know!

@stylelint/core What are your thoughts? Do you agree we need to grow the team? If so, got any ideas, other than asking the people listed above if they are interested, about how we could do this e.g. can our processes be improved somehow, can we lower the barrier to contributing, would a regular release cycle help etc?

**Comment by modosc** [comment]
i'd be happy to help

**Comment by hudochenkov** [comment]
> I think growing the stylelint team would be healthy and would help the sustainability of the project.

It's a good idea! Candidates are also good.

> I'm sure there are a few more suitable people I've failed to mention.

We could ask @Arcanemagus also. He's helping us with discussions on issues, and he's [AtomLinter/linter-stylelint](https://github.com/AtomLinter/linter-stylelint/) maintainer.

> I'll try to knock-up, in the next couple of days, additions to the developer guide to formalise the first two of these. We have an informal approach to issues and PRs that isn't documented anywhere

When I started as a stylelint maintainer you gave me these instructions:

* https://github.com/stylelint/stylelint/issues/1663#issuecomment-274260021
* https://github.com/stylelint/stylelint/pull/2275#issuecomment-274321815

I guess they could be a good starting point for developers guide update.

> Can our processes be improved somehow?

I think our processes are good. They are proven to keep sustainability and quality of stylelint, and its code base. The only thing we're missing is a time. But it's going to be better if we have a little more people on board. And I about to have more time to spend on stylelint development.

**Comment by Arcanemagus** [comment]
Thanks for the offer @hudochenkov, but I barely have time to try to keep up with all the repos in the AtomLinter org 😉. (Nevermind the fact that I know virtually nothing about CSS / PostCSS).

Feel free to tag me in issues or hit me up on [Atom's Slack](https://discuss.atom.io/t/join-us-on-slack/16638) if there's something you would like input on from the perspective of an API consumer 😉.

**Comment by gucong3000** [comment]
stylelint is cool, I like it. I'd be happy to join.

**Comment by CAYdenberg** [comment]
I'd like to join as well, if the team agrees that all three of us are needed/wanted. No pressure either way :smiley: 

My one question would be that if stylelint is growing to now 12(?) people with write access is there a need to discuss if specific people should be in charge of watching issues, providing code reviews etc..

**Comment by lvliang271486017** [comment]
i want you and i like to join in!

**Comment by jeddy3** [comment]
@CAYdenberg, @gucong3000 and @modosc Fantastic news that you're all interested!

@stylelint/core If there aren't any objections, and now that the pull requests and issue management guides [are merged](https://github.com/stylelint/stylelint/pull/2868), I'm going to add them to the team.

(@Arcanemagus Totally understand. Thanks, as always, for your continued work on AtomLinter. Especially, the linter-stylelint package! :))

> My one question would be that if stylelint is growing to now 12(?) people with write access is there a need to discuss if specific people should be in charge of watching issues, providing code reviews etc..

Good question. Only a handful of the existing 11 users are active, so there probably isn't enough scope, even with three new additions, to divide out the responsibilities. In the past, it's been informal where people have dipped into what they like and/or what they thought needed doing. We could continue like that for a while to see how things pan out and reassess it (once everyone has had time to settle in) if we think some formality would help. So, feel free to jump into the pull request list, issue backlog etc as you see fit for now :)

> i want you and i like to join in!

@lvliang271486017 Thanks for offering to help out! There is a [list of ways](https://github.com/stylelint/stylelint#help-out) that you can help out and start contributing to the project. Good luck :)

**Comment by lvliang271486017** [comment]
@jeddy3 thanks

**Comment by hudochenkov** [comment]
No objections from me :)

There is one thing that we forget to add to [docs/developer-guide/pull-requests.md](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/pull-requests.md). Never push to master. If you need to change something — create PR. Even if it's one letter change.

Also, I would like to ask new members, to have at least one older member approval on PRs at the beginning. To ensure new members are on track with our processes and vision.

**Comment by gucong3000** [comment]
> Never push to master. If you need to change something — create PR.
We can disable this permission.

**Comment by gucong3000** [comment]
> Also, I would like to ask new members, to have at least one older member approval on PRs at the beginning. To ensure new members are on track with our processes and vision.

No problem for me, I'd love to.

**Comment by jeddy3** [comment]
> There is one thing that we forget to add to docs/developer-guide/pull-requests.md. Never push to master. If you need to change something — create PR, even if it's a one letter change.

Sounds good. I'll expand out the [Prerequisites](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/prerequisites.md) section of the Developer Guide with this type of information. The pull request doc is specifically for _reviewing_ them.

> We can disable this permission in here: 

Unfortunately, we can't disable it outright as we commit directly to master (via the website) [when updating](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/pull-requests.md) the CHANGELOG. I'm hesitant to change this process as it fixed issues we had with updating the CHANGELOG in the past.

**Comment by Arcanemagus** [comment]
@gucong3000 Note that those options you show only prevent _force_ pushes, not regular commits, however the options you aren't showing (requiring CI checks) do prevent direct commits.

**Comment by ntwb** [comment]
No objections from me, everything above is awesome 💯

**Comment by jeddy3** [comment]
@CAYdenberg, @gucong3000 and @modosc I've sent you invitations to the stylelint organisation. Welcome onboard! :)

Please replace your local forks with a clone of this repo. This will make collaborating on (and rebasing) each other's PRs easier.

As I said above:

- You'll be able to help review pull requests, merge them and update the [CHANGELOG](https://github.com/stylelint/stylelint/blob/master/CHANGELOG.md).
- You'll be able to help triage issues.
- You'll be able to help put out releases.
- You'll be able to help decide the direction that stylelint takes going forward (see: [VISION.md](https://github.com/stylelint/stylelint/blob/master/VISION.md)).

Please familiarise yourselves with the [Developer Guide](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide.md), as it'll help you with the above. In particular:

- [The contributor guides](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide.md#contributor-guides) i.e. working on rules.
- [Core team guides](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide.md#core-team-guides) i.e. issues, pull requests and releases.

Additionally, it's probably worth reading through the [User guide](https://github.com/stylelint/stylelint/blob/master/docs/user-guide.md) too. In particular, [how rules work together](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/about-rules.md).

Don't hesitate to shout if you've any questions about, or have feedback on, these guides.

Feel free to jump into any of the issues or pull requests. Generally, we wait on external contributors to add the [new rules](https://github.com/stylelint/stylelint/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20label%3A%22type%3A%20new%20rule%22%20) and/or [new options](https://github.com/stylelint/stylelint/issues?q=is%3Aissue+is%3Aopen+label%3A%22type%3A+new+option%22) they have proposed (so that they have the opportunity and time to get involved). But you can jump on them if it's something that you'll use yourself. More importantly, there are a few key areas that the stylelint team are trying to push forward:

- Making the linting engine more robust e.g. https://github.com/stylelint/stylelint/issues/2833 and https://github.com/stylelint/stylelint/issues/2757
- Growing the new autofix system and making it more robust e.g. https://github.com/stylelint/stylelint/issues/2710, https://github.com/stylelint/stylelint/issues/2643 and https://github.com/stylelint/stylelint/issues/2829
- Addressing bugs e.g. [bugs label](https://github.com/stylelint/stylelint/issues?q=is%3Aissue+is%3Aopen+label%3A%22type%3A+bug%22)
- Improving the performance of stylelint e.g. https://github.com/stylelint/stylelint/issues/2454
- Shoring up our [flow](https://github.com/facebook/flow) coverage (if anyone of you has experience with, or fancies learning, flow then please jump in!) e.g. https://github.com/stylelint/stylelint/pull/2823 and https://github.com/stylelint/stylelint/pull/2714
- Empowering the ecosystem by improving our points of extensibility e.g.  https://github.com/stylelint/stylelint/issues/2502
- Improving our tests e.g. [test label](https://github.com/stylelint/stylelint/issues?q=is%3Aissue+is%3Aopen+label%3A%22type%3A+tests%22)

Right, I think that's enough of an introduction to stylelint. Hopefully it has provided a good overview.

Feel free to use this thread for questions.

Welcome onboard again!  :)

**Comment by hudochenkov** [comment]
Welcome, people! :)

**Comment by gucong3000** [comment]
![image](https://user-images.githubusercontent.com/1171691/30317925-30f5b614-97de-11e7-8233-eedbc50a388d.png)
@jeddy3 I didn't get an invitation.

**Comment by jeddy3** [comment]
> @jeddy3 I didn't get an invitation.

Weird. Can you try again, please? I've updated the invite.

**Comment by m-allanson** [comment]
This all sounds great 👍 Welcome, everyone!

**Comment by gucong3000** [comment]
I still haven't received the invitation. Maybe the username is wrong, it should be `gucong3000`

**Comment by jeddy3** [comment]
> I still haven't received the invitation. Maybe the username is wrong, it should be gucong3000

Strange. I've revoke the invitation and tried another from scratch. It's definitely the right username as your avatar is showing up :)

**Comment by ntwb** [comment]
The invitation should link to the _GitHub Orginization_ rather than this repo I believe:

e.g: Something along the lines of: `https://github.com/orgs/stylelint/invitation?via_email=1`

**Comment by gucong3000** [comment]
![image](https://user-images.githubusercontent.com/1171691/30319779-5f7a080a-97e3-11e7-90fd-0823080ca500.png)
@jeddy3 @ntwb All right now. Thank you.

**Comment by ntwb** [comment]
Congrats and welcome y'all :)

**Comment by CAYdenberg** [comment]
Thanks, everyone. Flow is a bit of a current interest of mine so I'll have a look at those two issues sometime soon.

**Comment by CAYdenberg** [comment]
So - I'd like to get some of the "stale" pull requests merged, especially those where I like what the author did but the tests are failing for whatever reason or there are issues that need to be addressed. Do we have a policy on just jumping in on other peoples PRs? I know other OS communities require you to branch from a branch and either submit a separate PR to master or a PR to the branch. What's good practice here?

**Comment by jeddy3** [comment]
> So - I'd like to get some of the "stale" pull requests merged

That would be amazing!

> Do we have a policy on just jumping in on other peoples PRs?

There's no offical policy (if only because we've been such a small team that nearly everything has, so far, been informal conventions). For other team member's PRs (that is branches off the main repo) we've generally added a comment to the PR saying you'll pick it up and then commit directly into that branch. For external contributors (working from a fork) it's a bit tricker and, I think, in the past we've just duped the work into a new PR.

Does this sound OK? More than happy to establish some other approach if you think it'll be benefical. I believe continuing to improve our processes will help keep the project sustainable.

**Comment by ntwb** [comment]
> _"For external contributors (working from a fork) it's a bit tricker and, I think, in the past we've just duped the work into a new PR."_

Depending on the contributors pull request you may also be able to push directly to that PR

https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

It's a pretty neat feature, my anecdotal analysis appears to be a 50% hit and miss, some contributors allow it, some do not.

> _"... we've generally added a comment to the PR saying you'll pick it up and then commit directly into that branch."_

Per the above, I'd suggest adding a comment to any PR you're planning on picking up regardless.

I'm not sure there is a way to tell if you are able to push to a contributors pull request without actually trying it, so that's what I do, I'll try adding a commit to the checked out PR and follow that with a `git push`, if it doesn't succeed I'll just create a fresh branch locally and copy that checkout to the new branch and go from there.

**Comment by jeddy3** [comment]
> Depending on the contributors pull request you may also be able to push directly to that PR

I didn't know about that. That's is neat!

**Comment by jeddy3** [comment]
I'm going to close this now as I think everyone is settled in.

This also seems like a good time to mention that stylelint just passed 1 million npm downloads a month. Congratulations everyone :) 🎉

**Comment by CAYdenberg** [comment]
Hey all, I've been a stranger lately as some contract work has come up, and also tightening my portfolio as I'm moving toward contracting fulltime sometime soon. Once I do leave my regular position I imagine my work level will ebb and flow somewhat.

**Comment by ntwb** [comment]
Thanks for the update @CAYdenberg, we've all been in that same position, and will again no doubt, good luck on the job front 👍
